### PR TITLE
LA-212 Add activities to workgroup layer data and presentation

### PR DESCRIPTION
**Pull request by dab246**
https://user-images.githubusercontent.com/80730648/113260078-0a764580-92f8-11eb-8c39-26be279b257a.mp4

### CODE REVIEW

File: data/lib/src/datasource_impl/shared_space_activities_datasource_impl.dart [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            } else if (auditLogEntryUserDto is WorkGroupDocumentAuditLogEntryDto) {
+              return auditLogEntryUserDto.toWorkGroupDocumentAuditLogEntry();
+            } else if (auditLogEntryUserDto is WorkGroupDocumentRevisionAuditLogEntryDto) {
+              return auditLogEntryUserDto.toWorkGroupDocumentRevisionAuditLogEntry();
+            } else if (auditLogEntryUserDto is WorkGroupFolderAuditLogEntryDto) {
+              return auditLogEntryUserDto.toWorkGroupFolderAuditLogEntry();
+            }
+            return null;
+          })
+          .where((node) => node != null)
```
Comment (hoangdat) [inline data/lib/src/datasource_impl/shared_space_activities_datasource_impl.dart:61]: We don't have an operator `mapNotNull`?

File: lib/presentation/util/extensions/audit_log_entry_user_extension.dart [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  String getWorkGroupNodeIconPath(AppImagePaths imagePath, WorkGroupNode workGroupNode) {
+    if (workGroupNode is WorkGroupDocument) {
+      return workGroupNode.mediaType.getFileTypeImagePath(imagePath);
+    }
+    return imagePath.icFolder;
+  }
+
+  String getResourceName(AuditLogEntryUser  auditLogEntryUser) {
```
Comment (hoangdat) [inline lib/presentation/util/extensions/audit_log_entry_user_extension.dart:75]: abstract functions for super class

File: domain/lib/src/model/audit/audit_log_entry.dart [hunk 3]
```diff
@@ -66,4 +66,6 @@ abstract class AuditLogEntry with EquatableMixin {
   ];
 
   Map<AuditLogActionMessageParam, dynamic> getActionMessageComponents();
+
+  String getResourceName();
```
Comment (hoangdat) [inline domain/lib/src/model/audit/audit_log_entry.dart:70]: Why not?
```
String getResourceName() {
    return resource.name;
}
```

File: data/lib/src/datasource_impl/shared_space_activities_datasource_impl.dart [hunk 4]
```diff
@@ -58,7 +58,6 @@ class SharedSpaceActivitiesDataSourceImpl implements SharedSpaceActivitiesDataSo
             }
             return null;
           })
-          .where((node) => node != null)
```
Comment (hoangdat) [inline data/lib/src/datasource_impl/shared_space_activities_datasource_impl.dart]: why you remove that?

File: domain/lib/src/model/audit/audit_log_entry.dart (hunk 3 above)
Comment (dab246) [reply domain/lib/src/model/audit/audit_log_entry.dart:70]: AuditLogEntry have't resource object

File: data/lib/src/datasource_impl/shared_space_activities_datasource_impl.dart (hunk 4 above)
Comment (dab246) [reply data/lib/src/datasource_impl/shared_space_activities_datasource_impl.dart]: It is not required for a map

**Review by hoangdat — APPROVED** [review]

File: data/lib/src/network/model/shared_space_activities/work_group_copy_dto.dart [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import 'package:data/src/network/model/converter/work_group_node_id_converter.dart';
+import 'package:data/src/util/attribute.dart';
+import 'package:domain/domain.dart';
+import 'package:equatable/equatable.dart';
+import 'package:json_annotation/json_annotation.dart';
+
+part 'work_group_copy_dto.g.dart';
+
+@JsonSerializable()
+@DatetimeConverter()
```
Comment (manhduc-nguyen) [inline data/lib/src/network/model/shared_space_activities/work_group_copy_dto.dart:43]: I don't think we need to use this

File: lib/presentation/widget/shared_space_details/shared_space_details_widget.dart [hunk 6]
```diff
@@ -258,4 +267,71 @@ class _SharedSpaceDetailsWidgetState extends State<SharedSpaceDetailsWidget> {
...
+    return Padding(
+        padding: EdgeInsets.only(top: 8),
+        child: RichText(
+            text: TextSpan(
+              style: TextStyle(fontSize: 13, color: AppColor.documentNameItemTextColor),
+              children: <TextSpan>[
+                TextSpan(
+                    text: actionDetails[AuditLogActionDetail.TITLE], style: TextStyle(fontWeight: FontWeight.bold, fontSize: 13, color: AppColor.documentNameItemTextColor)),
+                TextSpan(
+                    text: ' ' + actionDetails[AuditLogActionDetail.DETAIL],
```
Comment (manhduc-nguyen) [inline lib/presentation/widget/shared_space_details/shared_space_details_widget.dart:330]: I think its more readable if you write like this `text: ' ${actionDetails[AuditLogActionDetail.DETAIL]}' ` . WDYT

**Review by manhduc-nguyen — APPROVED** [review]
LGTM!

File: lib/presentation/widget/shared_space_details/shared_space_details_widget.dart [hunk 7]
```diff
@@ -258,4 +267,71 @@ class _SharedSpaceDetailsWidgetState extends State<SharedSpaceDetailsWidget> {
       }
     );
   }
-}
+
+  Widget _activitiesTabWidget(List<AuditLogEntryUser> activitiesList) {
+    return activitiesList.isEmpty ?
+      SizedBox.shrink() :
```
Comment (huynguyennovem) [inline lib/presentation/widget/shared_space_details/shared_space_details_widget.dart:273]: Recommend to use `const SizedBox.shrink()` here.

**Review by huynguyennovem — COMMENTED** [review]
/lgtm

File: lib/presentation/util/extensions/log_action_extension.dart [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+// more details.
+// You should have received a copy of the GNU Affero General Public License and its
+// applicable Additional Terms for LinShare along with this program. If not, see
+// <http://www.gnu.org/licenses/> for the GNU Affero General Public License version
+//  3 and <http://www.linshare.org/licenses/LinShare-License_AfferoGPL-v3.pdf> for
+//  the Additional Terms applicable to LinShare software.
+
+import 'package:domain/domain.dart';
+
+extension LogActionExtension on LogAction {
```
Comment (hoangdat) [inline lib/presentation/util/extensions/log_action_extension.dart:34]: How about with `action` is COPY in Vn, Fr, Ru?
I think it is better if you get the name in `presentation` to easier to localize this name with `app_localization`.

File: data/lib/src/network/model/shared_space_activities/work_group_copy_dto.dart [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+import 'package:data/src/network/model/converter/copy_context_id_converter.dart';
+import 'package:data/src/network/model/converter/datetime_converter.dart';
+import 'package:data/src/network/model/converter/work_group_node_id_converter.dart';
+import 'package:data/src/util/attribute.dart';
+import 'package:domain/domain.dart';
+import 'package:equatable/equatable.dart';
+import 'package:json_annotation/json_annotation.dart';
+
+part 'work_group_copy_dto.g.dart';
```
Comment (hoangdat) [inline data/lib/src/network/model/shared_space_activities/work_group_copy_dto.dart:40]: Can you share to me about that `part`

File: lib/presentation/util/extensions/log_action_extension.dart (hunk 8 above)
Comment (dab246) [reply lib/presentation/util/extensions/log_action_extension.dart:34]: This is not a value to show to the user. It is just identifier keywords. So I don't think there is a need to translate into other languages.

File: data/lib/src/network/model/shared_space_activities/work_group_copy_dto.dart (hunk 9 above)
Comment (dab246) [reply data/lib/src/network/model/shared_space_activities/work_group_copy_dto.dart:40]: This is the workgroup that has been copied by the user

Comment (dab246) [reply data/lib/src/network/model/shared_space_activities/work_group_copy_dto.dart:40]: `part` Where is it?

**Pull request merged**